### PR TITLE
Pagination, search, sorting for the Stage page.

### PR DESCRIPTION
**Pull request by danielgafni**
Providing `examples/1000tasks.py` to check out the new Stage page for a lot of tasks.

### CODE REVIEW

File: cosmos/web/views.py [hunk 1]
```diff
@@ -57,23 +58,126 @@ def workflow(name):
...
+                             workflow_name=workflow_name, stage_name=stage_name)
+
+        return render_template("cosmos/stage.html", stage=stage, drm_statuses=drm_statuses,
+                               tasks_on_page=tasks_paginated, max_page=max_page, colnames=colnames,
+                               ordering_urls=ordering_urls, page=page, url_handle=url_handle,
+                               first_url=first_url, prev_url=prev_url, next_url=next_url, last_url=last_url,
+                               workflow_name=workflow_name, stage_name=stage_name, keyword=keyword)
         # x=filter(lambda t: t.status == TaskStatus.submitted, stage.tasks))
 
+    @bprint.route("/workflow/<workflow_name>/<stage_name>/handle_stage", methods=['GET', 'POST'])
```
Comment (egafni) [inline cosmos/web/views.py:149]: "handle_stage" is very ambiguous, maybe rename to "query"?

Comment (egafni) [reply cosmos/web/views.py:149]: or "paginate"? what do you think is clearer

File: cosmos/web/views.py [hunk 2]
```diff
@@ -57,23 +58,126 @@ def workflow(name):
...
+        else:
+            keyword = old_keyword
+
+        if request.form.get('clear_search') == 'Clear':
+            keyword = None
+            sorting = None
+            order = None
+            page = 1
+
+        colnames = ["id", "task", "successful", "status", "drm_status", "drm_jobid",
```
Comment (egafni) [inline cosmos/web/views.py:173]: doesnt look like colnames is used

File: cosmos/web/templates/cosmos/stage.html [hunk 3]
```diff
@@ -70,30 +70,57 @@
...
+            <input type="submit" name="submit_search" value="Search">
+        </form>
+        <form action="{{ url_handle }}" method="post" style="display: inline">
+            {% if keyword not in ['', None] %}
+            <input type="submit" name="clear_search" value="Clear">
+            {% endif %}
+        </form>
+    </div>
+    <div style="margin-bottom: 0.5cm; margin-top: 0.5cm">
+        {% if first_url %}
```
Comment (egafni) [inline cosmos/web/templates/cosmos/stage.html:89]: turn to buttons
i think all you need is class="button" or class="button-small"

File: cosmos/web/search.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+from flask import current_app
```
Comment (egafni) [inline cosmos/web/search.py:1]: is this code ever used?

File: cosmos/models/Task.py [hunk 5]
```diff
@@ -196,7 +195,8 @@ class Task(Base):
     """
     # FIXME causes a problem with mysql?
     __table_args__ = (UniqueConstraint("stage_id", "uid", name="_uc1"),)
-
+    __searchable__ = ["id", "successful", "params",
+                      "attempt", "submitted_on", "finished_on", "wall_time"]
```
Comment (egafni) [inline cosmos/models/Task.py:199]: status needs to be searchable

File: cosmos/web/views.py [hunk 6]
```diff
@@ -57,23 +58,126 @@ def workflow(name):
         workflow = session.query(Workflow).filter_by(name=name).one()
         return render_template("cosmos/workflow.html", workflow=workflow)
 
-    @bprint.route("/workflow/<workflow_name>/<stage_name>/")
+    @bprint.route("/workflow/<workflow_name>/<stage_name>/", methods=['GET', 'POST'])
     def stage(workflow_name, stage_name):
+        IN_PAGE = 40  # how many tasks to show on one page
+        # these are the column names that appear in the screen
+        colnames = ["id", "task", "successful", "status", "drm_status", "drm_jobid",
```
Comment (egafni) [inline cosmos/web/views.py:65]: status and drm job id must be searchable

File: cosmos/web/views.py [hunk 7]
```diff
@@ -57,23 +58,126 @@ def workflow(name):
...
+        else:
+            pattern = keyword.replace("'", "''")
+            tasks_searched = tasks.filter(or_(
+                *[getattr(Task, field).like(f"%{keyword}%") if field != 'params' else None
+                  for field in Task.__searchable__], text(f"params LIKE '%{pattern}%'")
+                ))
+
+        # sort
+        tasks_sorted = tasks_searched
+        if sorting is not None:
```
Comment (egafni) [inline cosmos/web/views.py:94]: i should have been clearer, but just having one sort (either ascending or descending of a single column) would be sufficient

File: cosmos/web/views.py (hunk 1 above)
Comment (egafni) [reply cosmos/web/views.py:149]: its not supe robvious to me why you need a second function who's only purpose is to redirect.  if you think its necessary, can you write a comment why?

File: cosmos/web/views.py [hunk 8]
```diff
@@ -57,23 +58,126 @@ def workflow(name):
...
+        submitted = [t for t in tasks_paginated if t.status == TaskStatus.submitted]
         jm = JobManager(get_submit_args=None, logger=None)
 
         f = attrgetter("drm")
         drm_statuses = {}
         for drm, tasks in it.groupby(sorted(submitted, key=f), f):
             drm_statuses.update(jm.get_drm(drm).drm_statuses(list(tasks)))
 
-        return render_template("cosmos/stage.html", stage=stage, drm_statuses=drm_statuses)
+        url_handle = url_for('cosmos.handle_stage', old_page=page, old_keyword=keyword, sorting=sorting, order=order,
```
Comment (egafni) [inline cosmos/web/views.py:139]: "url_handle" is very ambiguous

File: cosmos/web/views.py (hunk 1 above)
Comment (danielgafni) [reply cosmos/web/views.py:149]: ok, changed to query

File: cosmos/web/search.py (hunk 4 above)
Comment (danielgafni) [reply cosmos/web/search.py]: removed it

File: cosmos/web/views.py (hunk 1 above)
Comment (danielgafni) [reply cosmos/web/views.py]: I use it to prevent resending the form on page refresh

File: cosmos/web/views.py (hunk 7 above)
Comment (danielgafni) [reply cosmos/web/views.py:93]: so should I remove the descending order?

Comment (egafni) [reply cosmos/web/views.py:93]: descending is still nice, thanks

File: examples/1000tasks.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        # example of how to change arguments if you're not using default_drm='local'
+        get_submit_args=partial(default_get_submit_args, parallel_env="smp"),
+        default_drm=args.drm,
+        default_max_attempts=2,
+        default_job_class=args.job_class,
+        default_queue=args.queue,
+    )
+    cosmos.initdb()
+
+    sp.check_call("mkdir -p analysis_output/ex2", shell=True)
```
Comment (egafni) [inline examples/1000tasks.py:52]: please update the folder name from ex2/ to 1000tasks/

File: cosmos/web/views.py [hunk 10]
```diff
@@ -57,23 +58,124 @@ def workflow(name):
...
+
+        return render_template("cosmos/stage.html", stage=stage, drm_statuses=drm_statuses, in_page=in_page,
+                               tasks_on_page=tasks_paginated, max_page=max_page, colnames=colnames,
+                               ordering_urls=ordering_urls, page=page, url_query=url_query,
+                               first_url=first_url, prev_url=prev_url, next_url=next_url, last_url=last_url,
+                               workflow_name=workflow_name, stage_name=stage_name, keyword=keyword)
         # x=filter(lambda t: t.status == TaskStatus.submitted, stage.tasks))
 
+    @bprint.route("/workflow/<workflow_name>/<stage_name>/query", methods=["GET", "POST"])
+    def query(workflow_name, stage_name):
```
Comment (egafni) [inline cosmos/web/views.py:151]: please rename to stage_query

File: cosmos/web/views.py [hunk 11]
```diff
@@ -57,23 +58,124 @@ def workflow(name):
...
+        page = request.args.get("old_page", 1, int)
+        keyword = request.args.get("old_keyword", "", type=str)
+        in_page = request.args.get("old_in_page", 40, type=int)
+        order = request.args.get("order", 1, int)
+        sorting = request.args.get("old_sorting", None, type=str)
+
+        if request.form.get("submit_page") == "Go to page":
+            page = request.form.get("page")
+
+        if request.form.get("submit_search") == "Search":
```
Comment (egafni) [inline cosmos/web/views.py:161]: it seems this should be
if
elif
elif..
else:
  raise AssertionError('invalid form')

**Pull request merged**